File: be-1-nestjs/src/auth/util.get-tokens.ts
import axios from 'axios';
import { GetTokensDto } from './auth.dtos';
import { config } from '@/config'

export interface GetTokensResponse {
  access_token: string,
  id_token: string,
  refresh_token: string,
  token_type: string, // Bearer Basic
  expires_in: number
}

export const getTokens = async ({ code, redirectUri }: GetTokensDto): Promise<GetTokensResponse> => {
  const { clientId, clientSecret, clientDomain, clientRegion } = config().cognito
  const url = `https://${clientDomain}.auth.${clientRegion}.amazoncognito.com/oauth2/token`
  const client_id = clientId
  const grant_type = 'authorization_code'
  const redirect_uri = redirectUri
  const params = { grant_type, client_id, code, redirect_uri }
  const Authorization = 'Basic ' + Buffer.from(client_id + ':' + clientSecret).toString('base64')

  const headers = {
    'Content-Type': 'application/x-www-form-urlencoded',
    'Authorization': Authorization
  }

  const response = await axios.request<GetTokensResponse>({ url, params, headers, method: 'POST' })
  return response.data
}

File: be-2-flask/app/auth/util_get_tokens.py
from typing import TypedDict
import requests
import base64
from config import (
    COGNITO_CLIENT_DOMAIN,
    COGNITO_CLIENT_ID,
    COGNITO_CLIENT_REGION,
    COGNITO_CLIENT_SECRET,
)


class GetTokensResponse(TypedDict):
    access_token: str
    id_token: str
    refresh_token: str
    token_type: str
    expires_in: int


async def get_tokens_util(code, redirect_uri) -> GetTokensResponse:
    url = f"https://{COGNITO_CLIENT_DOMAIN}.auth.{COGNITO_CLIENT_REGION}.amazoncognito.com/oauth2/token"
    client_id = COGNITO_CLIENT_ID
    client_secret = COGNITO_CLIENT_SECRET
    grant_type = "authorization_code"
    params = {"grant_type": grant_type, "client_id": client_id, "code": code, "redirect_uri": redirect_uri}

    authorization = "Basic " + base64.b64encode(f"{client_id}:{client_secret}".encode("utf-8")).decode("utf-8")

    headers = {"Content-Type": "application/x-www-form-urlencoded", "Authorization": authorization}

    print("...COGNITO --> getTokens")

    try:
        response = requests.post(url, data=params, headers=headers)
        response_data: GetTokensResponse = response.json()
        return response_data
    except Exception as e:
        raise Exception("COGNITO REQ ERROR")
